### PR TITLE
Revert "Revert "Validate function arguments in query tree""

### PR DESCRIPTION
**Pull request by kitaisreal**
Reverts ClickHouse/ClickHouse#45299

### CODE REVIEW

**Review by novikd — APPROVED** [review]

File: src/Analyzer/Passes/AggregateFunctionsArithmericOperationsPass.cpp [hunk 1]
```diff
@@ -47,19 +48,23 @@ Field zeroField(const Field & value)
 class AggregateFunctionsArithmericOperationsVisitor : public InDepthQueryTreeVisitor<AggregateFunctionsArithmericOperationsVisitor>
 {
 public:
+    explicit AggregateFunctionsArithmericOperationsVisitor(ContextPtr context_)
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Analyzer/Passes/AggregateFunctionsArithmericOperationsPass.cpp:51]: I will change all passes to proper use settings in https://github.com/ClickHouse/ClickHouse/pull/42976.

File: src/Analyzer/Passes/AggregateFunctionsArithmericOperationsPass.cpp [hunk 2]
```diff
@@ -69,88 +74,111 @@ class AggregateFunctionsArithmericOperationsVisitor : public InDepthQueryTreeVis
...
 
         /** If we extract negative constant, aggregate function name must be updated.
           *
           * Example: SELECT min(-1 * id);
           * Result: SELECT -1 * max(id);
           */
-        std::string function_name_if_constant_is_negative;
-        if (inner_function_name == "multiply" || inner_function_name == "divide")
+        std::string aggregate_function_name_if_constant_is_negative;
+        if (lower_aggregate_function_name == "multiply" || lower_aggregate_function_name == "divide")
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Analyzer/Passes/AggregateFunctionsArithmericOperationsPass.cpp:106]: Updated.

File: src/Analyzer/Passes/AggregateFunctionsArithmericOperationsPass.cpp [hunk 3]
```diff
@@ -69,88 +74,111 @@ class AggregateFunctionsArithmericOperationsVisitor : public InDepthQueryTreeVis
...
+                aggregate_function_name_if_constant_is_negative = "min";
         }
 
+        size_t arithmetic_function_argument_index = 0;
+
         if (left_argument_constant_node && !right_argument_constant_node)
         {
             /// Do not rewrite `sum(1/n)` with `sum(1) * div(1/n)` because of lose accuracy
-            if (inner_function_name == "divide")
+            if (lower_aggregate_function_name == "divide")
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Analyzer/Passes/AggregateFunctionsArithmericOperationsPass.cpp:119]: Updated.

File: src/Analyzer/Passes/AggregateFunctionsArithmericOperationsPass.cpp [hunk 4]
```diff
@@ -69,88 +74,111 @@ class AggregateFunctionsArithmericOperationsVisitor : public InDepthQueryTreeVis
...
+        const std::string & result_aggregate_function_name)
+    {
+        auto arithmetic_function_clone = arithmetic_function->clone();
+        auto & arithmetic_function_clone_typed = arithmetic_function_clone->as<FunctionNode &>();
+        auto & arithmetic_function_clone_arguments_nodes = arithmetic_function_clone_typed.getArguments().getNodes();
+        auto & arithmetic_function_clone_argument = arithmetic_function_clone_arguments_nodes[arithmetic_function_argument_index];
+
+        auto aggregate_function_clone = aggregate_function->clone();
+        auto & aggregate_function_clone_typed = aggregate_function_clone->as<FunctionNode &>();
+        aggregate_function_clone_typed.getArguments().getNodes() = {arithmetic_function_clone_argument};
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Analyzer/Passes/AggregateFunctionsArithmericOperationsPass.cpp:167]: Updated.

**Pull request merged**